Buildings/Controls/OBC/Utilities/PIDWithAutotuning/Relay/HalfPeriodRatio.mo

```diff
@@ -0,0 +1,182 @@
+          -42,-24}},            color={0,0,127}));
+  connect(AddtOntOff.u1, tOn) annotation (Line(points={{-82,-24},{-94,-24},{-94,
+          60},{-120,60}},                   color={0,0,127}));
+  connect(AddtOntOff.y, mul.u2) annotation (Line(points={{-58,-30},{-50,-30},{-50,
+          -36},{-42,-36}},                     color={0,0,127}));
+  connect(mul.y, gretOntOff.u1) annotation (Line(points={{-18,-30},{-8,-30},{-8,
+          30},{-2,30}}, color={0,0,127}));
+  connect(mul.y, tIncrease.u1)
+    annotation (Line(points={{-18,-30},{38,-30}}, color={0,0,127}));
+  connect(gretmaxtOntOff.y, triSta)
+    annotation (Line(points={{22,0},{120,0}}, color={255,0,255}));
+  connect(gretmaxtOntOff.u2, gretOntOff.u2) annotation (Line(points={{-2,8},{
+          -16,8},{-16,30},{-10,30},{-10,38},{-2,38}}, color={0,0,127}));
+  connect(gretmaxtOntOff.u1, mul.u2) annotation (Line(points={{-2,0},{-50,0},{
+          -50,-36},{-42,-36}}, color={0,0,127}));
+  connect(tDecrease.u1, tIncrease.u1) annotation (Line(points={{38,-80},{34,-80},
+          {34,-30},{38,-30}}, color={0,0,127}));
+  connect(tDecrease.u2, tIncrease.u2) annotation (Line(points={{38,-88},{28,-88},
+          {28,-38},{38,-38}}, color={0,0,127}));
+  connect(tChanges.u1, tIncrease.y)
+    annotation (Line(points={{68,-30},{62,-30}}, color={255,0,255}));
+  connect(tChanges.u2, tDecrease.y)
+    annotation (Line(points={{68,-38},{68,-80},{62,-80}}, color={255,0,255}));
+  connect(tChanges.y, tOffSample.trigger) annotation (Line(points={{92,-30},{92,
+          -60},{26,-60},{26,-52},{-70,-52},{-70,-58}}, color={255,0,255}));
+  annotation (
+        Diagram(
+           coordinateSystem(
+           extent={{-100,-100},{100,100}})),
+        defaultComponentName = "halPerRat",
+        Icon(coordinateSystem(preserveAspectRatio=false, extent={{-100,-100},{80,
+            100}}),                                      graphics={
+        Rectangle(
+          extent={{-100,-100},{100,100}},
+          lineColor={0,0,127},
+          fillColor={255,255,255},
+          fillPattern=FillPattern.Solid),
+        Text(
+          extent={{-154,148},{146,108}},
+          textString="%name",
+          textColor={0,0,255})}),
+    Documentation(revisions="<html>
+<ul>
+<li>
+June 1, 2022, by Sen Huang:<br/>
+First implementation<br/>
+</li>
+</ul>
+</html>", info="<html>
+<p>This block calculates the half-period ratio of the output from a relay controller.</p>
+<h4>Main equations</h4>
+<p align=\"center\" style=\"font-style:italic;\">
+&rho; = max(t<sub>on</sub>,t<sub>off</sub>)/ min(t<sub>on</sub>,t<sub>off</sub>),
+</p>
+<p>where <i>t<sub>on</sub></i> and <i>t<sub>on</sub></i> are the length of the On period and the Off period, respectively.</p>
+<p>During an On period, the relay switch signal becomes True;</p>
+<p>During an Off period,  the relay switch signal becomes False.</p>
+<p>Note that only the first On period and the first Off period are considered.</p>
```

Comment:
The code implementation implement more than describe here, like how to trigger the tuning start/end. Need a better documentation.

Reply:
I add an <code>algorithm</code> section, please let me know if more explanations are needed.